src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+using Microsoft.CodeAnalysis.LanguageServer.Handler.DebugConfiguration;
+using Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Threading;
+using Microsoft.CodeAnalysis.Workspaces.ProjectSystem;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.Extensions.Logging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+/// <summary>
+/// Workspace which handles file-based programs ('dotnet run app.cs' scenario).
+/// </summary>
+internal sealed class LspFileBasedProgramWorkspace : Workspace, ILspService, ILspWorkspace, ILspFileBasedProgramWorkspace
+{
+    /// <summary>
+    /// A single gate for code that is adding work to <see cref="_projectsToLoadAndReload" />. This is just we don't have code simultaneously trying to load and unload solutions at once.
+    /// </summary>
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+    // TODO: should we use the LanguageServerProjectSystem here instead?
+    private readonly AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)> _projectsToLoadAndReload; // TODO: do we need to introduce a shutdown cancellation token for this?
+
+    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly ILspServices _lspServices;
+    private readonly IFileChangeWatcher _fileChangeWatcher;
+    private readonly IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> _dynamicFileInfoProviders;
+    private readonly ProjectTargetFrameworkManager _projectTargetFrameworkManager;
+    private readonly ILoggerFactory _loggerFactory;
+
+    public bool SupportsMutation => true;
+
+    // ILspLogger rationalize
+    ILogger iLogger;
+
+    public LspFileBasedProgramWorkspace(
+        ILspServices lspServices,
+        IFileChangeWatcher fileChangeWatcher,
+        [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> dynamicFileInfoProviders,
+        ProjectTargetFrameworkManager projectTargetFrameworkManager,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        HostServices hostServices) : base(hostServices, WorkspaceKind.FileBasedPrograms)
+    {
+        _lspServices = lspServices;
+        _fileChangeWatcher = fileChangeWatcher;
+        _dynamicFileInfoProviders = dynamicFileInfoProviders;
+        _projectTargetFrameworkManager = projectTargetFrameworkManager;
+        _loggerFactory = loggerFactory;
+        _projectsToLoadAndReload = new AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)>(
+            TimeSpan.FromMilliseconds(100),
```

Comment:
use DelayConstants.    

Do we need to do this 10 times a second?  Is this to handle files that change on disk?  If so, generally our response to IO changes is 'Medium' (which is defined as twice a second).

Reply:
This came from the existing project system project -- I'm guessing the short time span was chosen because this queue is used for both the initial load and reload, thus a regular load is implicitly slowed down by this. Agreed in the case of a reload though we should be a bit slower.